src/coreclr/src/tools/Common/TypeSystem/Common/CastingHelper.cs

```diff
@@ -20,6 +20,79 @@ public static bool CanCastTo(this TypeDesc thisType, TypeDesc otherType)
             return thisType.CanCastToInternal(otherType, null);
         }
 
+        /// <summary>
+        /// Get TypeFlags of the reduced type of a type.
+        /// The reduced type concept is described in ECMA 335 chapter I.8.7
+        /// </summary>
+        static TypeFlags GetReducedTypeElementType(TypeDesc type)
+        {
+            TypeFlags elemType = type.GetTypeFlags(TypeFlags.CategoryMask);
+            switch (elemType)
+            {
+                case TypeFlags.Byte:
+                    return TypeFlags.SByte;
+                case TypeFlags.UInt16:
+                    return TypeFlags.Int16;
+                case TypeFlags.UInt32:
+                    return TypeFlags.Int32;
+                case TypeFlags.UInt64:
+                    return TypeFlags.Int64;
+                case TypeFlags.UIntPtr:
+                    return TypeFlags.IntPtr;
+            }
+
+            return elemType;
+        }
+
+        /// <summary>
+        /// Get CorElementType of the verification type of a type.
+        /// The verification type concepts is described in ECMA 335 chapter I.8.7
+        /// </summary>
+        static TypeFlags GetVerificationTypeElementType(TypeDesc type)
+        {
+            TypeFlags reducedTypeElementType = GetReducedTypeElementType(type);
+
+            switch (reducedTypeElementType)
+            {
+                case TypeFlags.Boolean:
+                    return TypeFlags.SByte;
+                case TypeFlags.Char:
+                    return TypeFlags.Int16;
+            }
+
+            return reducedTypeElementType;
+        }
+
+        /// <summary>
+        /// Check if verification types of two types are equal
+        /// </summary>
+        static bool AreVerificationTypesEqual(TypeDesc type1, TypeDesc type2)
+        {
+            if (type1 == type2)
+            {
+                return true;
+            }
+
+            TypeFlags e1 = GetVerificationTypeElementType(type1);
```

Comment:
Nit: inlining these temporaries into the `return` statement will speed things up here (no need to compute the verification element type for both if `type1.IsPrimitive` is false.

Reply:
Makes sense.